src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs

```diff
@@ -134,21 +133,63 @@ public HttpClient CreateClient(string name)
         }
 
         public HttpMessageHandler CreateHandler(string name)
+        {
+            return CreateHandler(name, null);
+        }
+
+        public HttpMessageHandler CreateHandler(string name, IServiceProvider services)
         {
             if (name == null)
             {
                 throw new ArgumentNullException(nameof(name));
             }
 
-            ActiveHandlerTrackingEntry entry = _activeHandlers.GetOrAdd(name, _entryFactory).Value;
+            ActiveHandlerTrackingEntry entry = _activeHandlers.GetOrAdd(
+                name,
+                (name) =>
+                {
+                    return new Lazy<ActiveHandlerTrackingEntry>(() =>
+                    {
+                        return CreateHandlerEntry(name, services);
+                    }, LazyThreadSafetyMode.ExecutionAndPublication);
+                }
+            ).Value;
 
             StartHandlerEntryTimer(entry);
 
-            return entry.Handler;
+            if (!entry.IsPrimary)
+            {
+                return entry.Handler;
+            }
```

Comment:
I'm probably not going to be able to give you the detailed review that this requires, but one of the things managing complexity here is that `_activeHandlers` is a single resource to synchronize access to.

Introducing another resource to keep in sync `_expiredHandlers` might dictate the use of a lock instead of a concurrent dictionary, because you now have to think about race conditions between enregistering data in these two collections.

Likewise, you're mutating `entry` below this comment - but `entry` is already in the concurrent dictionary and thus available for other threads to observe in an inconsistent state.

Reply:
Sure, thread safety definitely should (and will) be addressed with a lot more care than right now.

I don't see how I am mutating the entry though... 🤔 I believe I only read it, and BuildChain doesn't edit the handler from the entry either, it only assigns it to the end of the chain